### PR TITLE
WebAuthn support

### PR DESCRIPTION
**Pull request by jpaniagualaconich**
Implement a new second factor method that uses WebAuthn.

## Description

A new authentication method is available in the setup wizard: WebAuthn, that enables support for hardware authentication devices such as fido-u2f keys in modern browsers that implement the specification.
 
## Motivation and Context

WebAuthn is supported in the Chrome, Firefox, and Edge browsers, for credential creation and assertion using a U2F Token, like those provided by Yubico and Feitian.

## How Has This Been Tested?

New view and utils tests have been added.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by moggers87 — CHANGES_REQUESTED** [review]
I've only done a quick look over, but here are things that popped out to me right away:

- the new dependency should be optional, like the yubikey dependencies are
- a lot of the translation file updates appear to be unrelated to this PR and make it difficult to see if there's anything other than line number changes in there
- Is the client side JS really required? I'd prefer not to have any JS involved at all if possible. Also there are no tests for this non-trivial code.
- The `TWO_FACTOR_DEVICE_PREFERENCE` stuff looks interesting, but really needs its own PR

**Comment by moggers87** [comment]
By the looks of https://github.com/w3c/webauthn/issues/1255 it seems the JS is not optional, so that's a huge :-1: from me. Not enough for me to reject this PR out of hand, but that code really does need tests and the entirety of webauthn needs to be optional in this package. That JS also needs to be taken out of the templates so people who aren't using the default templates don't have to go write their own JS + tests.

**Comment by claudep** [comment]
It would be much easier to host such optional modules if we could finish and merge #331

**Comment by moggers87** [comment]
It's on my todo list!

**Comment by jpaniagualaconich** [comment]
Hi there. Shall I wait for #331 in that case? It isn't the best thing in the world that I "if cased" webauthn sprinkling it here and there especially in the SetupView. Mimicking what's been done for the yubikey case to make it optional would only go further down that road.

And no, JS code is not optional. Do you have a preference for the testing framework I should use?

**Comment by moggers87** [comment]
I have no preference on what testing framework you use, whatever you're already familiar with is probably best.

**Comment by jpaniagualaconich** [comment]
Hi there!

> * the new dependency should be optional, like the yubikey dependencies are

now it's optional

> * a lot of the translation file updates appear to be unrelated to this PR and make it difficult to see if there's anything other than line number changes in there

took out all the translation-related stuff

> * Is the client side JS really required? I'd prefer not to have any JS involved at all if possible. Also there are no tests for this non-trivial code.

changed it so JS is sent to the client only when WebAuthn is being used. Added an E2E test that covers the whole process (registration and use).
webauthn tests will be skipped if the webauthn package is not installed.

The E2E test will be skipped if:
- one of these packages is missing: webauthn, selenium, webdriver-manager. First one is an extra (`pip install django-two-factor-auth[webauthn]`). For the other two there's a `requirements_e2e.txt` file. 
- the `StaticLiveServerTestCase` service can't be reached at https://localhost/.
I added an `ssl-rev-proxy` make target that sets up the root CA and a localhost certificate needed to serve https.
It also spins an nginx container to reverse proxy from https://localhost.
It needs docker and mkcert.

The "webauthn is an extra" part is in the documentation but the "how to run the E2E test" part is not. I'd like to know what you think of the whole thing before attempting to document it.

> * The `TWO_FACTOR_DEVICE_PREFERENCE` stuff looks interesting, but really needs its own PR

it's no longer a part of this PR.

File: two_factor/templates/two_factor/core/login.html
Comment (moggers87) [inline two_factor/templates/two_factor/core/login.html:55]: There should be no JS in HTML like this. It makes implementing CSP difficult as you can no longer tell the browser to block inline JS.

File: two_factor/templates/two_factor/core/setup.html
Comment (moggers87) [inline two_factor/templates/two_factor/core/setup.html:60]: As above.

**Review by moggers87 — DISMISSED** [review]
I think we could add the e2e tests to tox.ini in a similar way to how yubikey tests were added.

File: two_factor/templates/two_factor/core/login.html
Comment (jpaniagualaconich) [reply two_factor/templates/two_factor/core/login.html]: No more inline JS

File: two_factor/templates/two_factor/core/setup.html
Comment (jpaniagualaconich) [reply two_factor/templates/two_factor/core/setup.html]: No more inline JS

**Comment by jpaniagualaconich** [comment]
- Did away with inline JS 
- No need for docker, mkcert, docker-ngin. django-sslserver is used instead
- Documentation is updated
- Chrome can silently stop cooperating if you forget some settings. Made it harder for this to happen.
- modified tox.ini to include webauthn testing, with E2E bits too

**Comment by brugnara** [comment]
Hi @moggers87 

Could you please help us and review this PR?

Thank you for considering my request.

**Comment by moggers87** [comment]
I'd like to wait for #425 to be merged before reviewing any more PRs

**Comment by jpaniagualaconich** [comment]
There's a new version of the `py_webauthn` package which brings support for lots of interesting attestation statement formats. Here is the complete list:
https://github.com/duo-labs/py_webauthn/releases/tag/v1.0.0

It does come with a totally new API, though, and now it requires python 3.8+.
@moggers87 since we're waiting for #425 I'm tempted to update. What do you recommend?

**Comment by Frikster** [comment]
For those waiting on this who don't want to use this development branch, I have tested and found both the following repos [django-fido](https://github.com/CZ-NIC/django-fido) and [kagi](https://github.com/justinmayer/kagi) to work quite nicely to offer Webauth 2FA.

**Comment by moggers87** [comment]
@jpaniagualaconich not sure I like bumping the minimum required version of Python to 3.8 just yet. 3.7 will still be receiving updates until June 2023: https://endoflife.date/python

**Comment by jpaniagualaconich** [comment]
@moggers87 ok, I patched `typing` so it has `Literal` also on python 3.7. This is what the new version of `webauthn` expects and it's the only reason why it wasn't working on 3.7.

I merged latest master and it's now ready for you to review.

**Review by moggers87 — COMMENTED** [review]
I got about half way through reviewing this and #331 was ready for merging. That's caused some conflicts.

How much work would it be to move webauthn specific code to the new plugins folder?

**Comment by claudep** [comment]
@jpaniagualaconich I'm still working on making this package more plugin-friendly, so some more "breakages" may come soon in master branch. I'm sorry for this, and will try to help as much as possible for your branch integration.

**Comment by jpaniagualaconich** [comment]
@claudep ok no problem.

@moggers87 ok, it uses django-extensions now and everything lives in `plugins/webauthn`.

**Comment by jpaniagualaconich** [comment]
@moggers87 @claudep I merged latest master and went with the _pluginisation_ as far as the current framework allows.

**Comment by moggers87** [comment]
@claudep I'll do a review, but I wouldn't mind a second pair of eyes to take a look.

**Comment by claudep** [comment]
Sure, that's my intention, but I'd like to finish integrating the email plugin first.

**Comment by hansegucker** [comment]
Is there any progress on this topic or something one could help with?

**Comment by jpaniagualaconich** [comment]
I merged latest master, again. It's ready for someone to review.

**Comment by claudep** [comment]
@jpaniagualaconich Do you think you would be able to extract the `backup_phones` replacement by get_devices/get_other_devices in a separate PR? (I may ask you other such separate PR's in the future, hope you don't mind).

**Comment by claudep** [comment]
Would it also possible to move the migration file inside plugins/webauthn dir?

**Comment by jpaniagualaconich** [comment]
@claudep I extracted the device/other devices thingy into its own PR: https://github.com/jazzband/django-two-factor-auth/pull/526

**Comment by jpaniagualaconich** [comment]
@claudep migration file moved, also webauthn tests moved to their own folder (`two_factor/plugins/webauthn/tests/`)
It's ready for another review round.

**Review by claudep — CHANGES_REQUESTED** [review]
What I'd still like to explore is the option to not push the JS files through the templates engine, but add the webauthn options through the data attributes of the form fields. This could however also be a follow-up optimization, if possible.
Regarding the correctness of the webauthn process, I trust you (and the tests).
Thanks a lot for your work and your patience.

File: Makefile
Comment (jpaniagualaconich) [reply Makefile:2]: yes, it is. If left untouched it will only look in the tests folder.

File: .gitignore
Comment (jpaniagualaconich) [reply .gitignore:14]: I sincerely wonder if changes such as this one really get in the way of understanding the rest of the PR.

File: setup.cfg
Comment (jpaniagualaconich) [reply setup.cfg:2]: Didn't feel like changing current behaviour which is to have flake8 ignore migrations. Perhaps this one merits its own PR.

File: .gitignore
Comment (WhyNotHugo) [reply .gitignore:14]: Hint: Put this kind of ignores in `$HOME/.config/git/ignore`. They'll apply to any repository on which you operate locally, and you won't need to add IDE-specific ignores to every single repo you work on.

File: Makefile
Comment (claudep) [reply Makefile:2]: Ah yes, then it's fine!

File: .gitignore
Comment (claudep) [reply .gitignore:14]: This discussion is now on #534.

For me, having pure self-contained changes with specific commits is important (I didn't say an obsession :smile: )

File: setup.cfg
Comment (claudep) [reply setup.cfg:2]: Absolutely, this would be for another PR (you begin to know me!)

File: .gitignore
Comment (jpaniagualaconich) [reply .gitignore]: excluded from PR as requested

File: docs/installation.rst
Comment (jpaniagualaconich) [reply docs/installation.rst]: All the other codeblocks in the same file were doing INSTALLED_APPS as a tuple. It's fine by me if INSTALLED_APPS is a list but in that case all of them should be, so that's what I did.

Comment (claudep) [reply docs/installation.rst]: Suggested as a separate PR in #538

File: setup.cfg
Comment (jpaniagualaconich) [reply setup.cfg]: i just made my only migration flake8-compatible to route around the controversy. Still think excluding migrations from flake8 checks is the way to go, though, as it's auto-generated code that's hardly ever modified by hand.

File: tests/settings.py
Comment (jpaniagualaconich) [reply tests/settings.py]: not needed. I took it out.

File: tests/test_views_setup.py
Comment (jpaniagualaconich) [reply tests/test_views_setup.py]: yes, that's where it should live now. Moved.

File: two_factor/plugins/webauthn/__init__.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/__init__.py]: gone.

File: two_factor/plugins/webauthn/admin.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/admin.py]: done.

File: two_factor/plugins/webauthn/admin.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/admin.py]: done.

File: two_factor/plugins/webauthn/apps.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/apps.py]: you're right. done.

File: two_factor/plugins/webauthn/views.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/views.py]: ok

File: two_factor/plugins/webauthn/templates/two_factor_webauthn/create_credential.js
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/templates/two_factor_webauthn/create_credential.js]: ok

File: two_factor/plugins/webauthn/tests/test_e2e.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/tests/test_e2e.py]: ok

File: two_factor/utils.py
Comment (jpaniagualaconich) [reply two_factor/utils.py:63]: ok, it's on #539 now.

File: two_factor/views/core.py
Comment (jpaniagualaconich) [reply two_factor/views/core.py:221]: ok: #541

File: two_factor/views/core.py
Comment (jpaniagualaconich) [reply two_factor/views/core.py:557]: ok: #541

File: two_factor/plugins/webauthn/forms.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/forms.py]: ok, made it more specific.

File: two_factor/plugins/webauthn/forms.py
Comment (jpaniagualaconich) [reply two_factor/plugins/webauthn/forms.py]: ok, made it more specific.

File: two_factor/templates/two_factor/core/login.html
Comment (jpaniagualaconich) [reply two_factor/templates/two_factor/core/login.html]: done.

File: CHANGELOG.md
Comment (jpaniagualaconich) [reply CHANGELOG.md]: right, done.

File: docs/configuration.rst
Comment (claudep) [reply docs/configuration.rst:165]: Wow, this is a nice and welcome improvement, great!

**Comment by claudep** [comment]
I guess we are approaching the merge moment, eventually! What do you think about rebasing and squashing the commits? Or we can do it during the merge, too.

**Comment by jpaniagualaconich** [comment]
@claudep great idea. done.

**Review by claudep — APPROVED** [review]

**Comment by claudep** [comment]
@moggers87, I think this is ready to go. can you give a last review or your approval?

**Comment by claudep** [comment]
Thanks a LOT @jpaniagualaconich for the constant rebasing and great patience to add this new functionality.

**Comment by jpaniagualaconich** [comment]
woohoo! 
thanks @claudep @moggers87 and all.

**Comment by Frikster** [comment]
Been lurking and waiting for this to merge for over a year, haha. Congrats on the merge.

**Comment by Be-ing** [comment]
Thank you @jpaniagualaconich, @moggers87, and @claudep!

**Comment by Crocmagnon** [comment]
🎉  That's great news! When can we expect a release including this feature?

**Pull request merged**
